include/fi_util.h

```diff
@@ -103,6 +104,47 @@ enum {
 };
 
 /*
+ * Utility classes/functions
+ */
+/*
+ * The key_idx object and related functions can be used to generate unique keys
+ * from an index. The key and index would refer to an object defined by the user.
+ * A local endpoint can exchange this key with a remote endpoint in the first message.
+ * The remote endpoint would then use this key in subsequent messages to reference
+ * the correct object at the local endpoint.
+ */
+struct ofi_key_idx {
+	uint64_t seq_no;
+	/* The uniqueness of the generated key would depend on how many bits are
+	 * used for the index */
+	uint8_t idx_bits;
+};
+
+static inline struct ofi_key_idx *ofi_key_idx_alloc(size_t size)
+{
+	struct ofi_key_idx *key_idx = calloc(1, sizeof(*key_idx));
+	if (!key_idx)
+		return NULL;
+	key_idx->idx_bits = fi_size_bits((uint64_t)size);
+	return key_idx;
+}
+
+static inline void ofi_key_idx_free(struct ofi_key_idx *key_idx)
+{
+	free(key_idx);
+}
+
+static inline uint64_t ofi_idx2key(struct ofi_key_idx *key_idx, uint64_t idx)
+{
+	return ((++(key_idx->seq_no)) << key_idx->idx_bits) | idx;
+}
+
+static inline uint64_t ofi_key2idx(struct ofi_key_idx *key_idx, uint64_t key)
+{
+	return key & ((1ULL << key_idx->idx_bits) - 1);
+}
```

Comment:
The connection map code can be moved into a separate header file. The key_idx abstraction would be used in places other than connection map as well so it could be left in fi_util.h. Is that fine?